xc/device.ml

```
@@ -1963,6 +1974,96 @@ module Dm = struct
+
+    let remove domid =
+      if Hashtbl.mem domid_to_connection_tbl domid
+      then begin
+        try
+          let c = Hashtbl.find domid_to_connection_tbl domid in
+          Hashtbl.remove domid_to_connection_tbl domid;
+          Hashtbl.remove connection_to_domid_tbl c;
+          Epoll.ctl epfd Epoll.EPOLL_CTL_DEL (Qmp_protocol.to_fd c) [Epoll.EPOLLIN];
+          Qmp_protocol.close c
+        with
+        | Unix.Unix_error(Unix.EBADF, "close", p) -> raise(Internal_error (Printf.sprintf "Failed to close qmp socket on domain-%d: %s" domid p))
+        | e -> raise(Internal_error (Printf.sprintf "Get error trying remove qmp on domain-%d: %s" domid (Printexc.to_string e)))
+      end
+    in
+
+    let add domid =
+      if not(Hashtbl.mem domid_to_connection_tbl domid)
+      then begin
+        let path = Printf.sprintf "/var/run/xen/qmp-event-libxl-%d" domid in
+        try
+          let c = Qmp_protocol.connect path in
+          Qmp_protocol.negotiate c;
+          Hashtbl.replace domid_to_connection_tbl domid c;
+          Hashtbl.replace connection_to_domid_tbl c domid;
+          Epoll.ctl epfd Epoll.EPOLL_CTL_ADD (Qmp_protocol.to_fd c) [Epoll.EPOLLIN];
+          debug "QMP domain-%d: negotiation complete" domid
+        with e ->
+          debug "QMP domain-%d: negotiation failed (%s): removing socket" domid (Printexc.to_string e);
+          remove domid
+      end
+    in
+
+    while true do
+      let cs = Hashtbl.fold (fun k v acc -> v :: acc) domid_to_connection_tbl [] in
+      let fds = List.map (fun x -> Qmp_protocol.to_fd x, x) cs in
+      let rs = Epoll.wait epfd (Hashtbl.length domid_to_connection_tbl + 1) (-1) in
+      List.iter (
+        fun ef ->
+          let events = fst ef in
+          let fd = snd ef in
+          if fd = qmp_event_pipe_r then begin
+            List.iter (
+              function
+              | Epoll.EPOLLIN ->
+                add (int_of_string (input_line qmp_event_channel_r))
```

can we ever receive a non-integer here? If there's an error here, we should print a debug message but not let the exception bubble up to kill the thread.

---

Added, thanks